File: westcoast/src/main/java/frc/vision/Targeting/StayOnTarget.java
package frc.vision.Targeting;

import edu.wpi.first.wpilibj.controller.PIDController;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.math.Maths;
import frc.vision.*;
import frc.robotMaps.RobotMap;

public class StayOnTarget extends CommandBase {
  private RobotMap map;
  private PIDController pid;
  private DifferentialDrive drive;
  private VisionRunner vision;
  private double value;

  public StayOnTarget(
    RobotMap map,
    DifferentialDrive drive,
    VisionRunner vision
  ) {
    this.map = map;
    this.drive = drive;
    this.vision = vision;
  }

  @Override
  public void initialize() {
    pid = map.makeStayOnTargetPIDs();
  }

  @Override
  public void execute() {
    value = vision.getAlignX();

    double turn = Maths.clamp(pid.calculate(value, 0), 0.4);

    drive.arcadeDrive(0, turn, false);
  }

  @Override
  public void end(boolean interrupted) {
    pid.reset();
    drive.stopMotor();
  }

  @Override
  public boolean isFinished() {
    return false;
  }
}

File: westcoast/src/main/kotlin/frc/commands/targeting/StayOnTarget.kt
package frc.commands.targeting

import edu.wpi.first.wpilibj.controller.PIDController
import edu.wpi.first.wpilibj2.command.PIDCommand
import frc.subsystems.DriveTrainSubsystem

class StayOnTarget(
    controller: PIDController,
    private val drive: DriveTrainSubsystem,
) : PIDCommand(
    controller,
    { drive.vision.alignX },
    0.0,
    { drive.arcade(0.0, it.coerceIn(-0.4, 0.4), false) },
    drive
) {

  override fun end(interrupted: Boolean) {
    drive.stop()
  }
}
